### PR TITLE
batch insertion events in archive database

### PR DESCRIPTION
**Pull request by ghost-not-in-the-shell**
Explain your changes:
This PR does batch insertion for `zkapp_events` in archive db.

Since `Account_update.Body.Events'.t` is defined as `field array list`, which is ismorphic to a list of list of fields.

We could batch the insertion of field and field_array to optimize the speed of archiving max-cost zkapps.

1. we flatten the list of list of fields to get all the field elements
2. insert all the field elements in one query
3. construct a map _M_ from `field_id` to `field` by querying against the zkapp_field table
4. use _M_ and the list of list of fields to compute the list of list of field_ids
5. insert all list of `list of field_ids` in one query
6. construct a map _M'_ from `field_array_id` to `field_id array` by querying against the zkapp_field_array table
7. use _M'_ and the list of list of field_ids to compute the list of field_array_ids
8. insert the list of field_arrays

To patch the existing database use `add_unique_constraints.sql`

Explain how you tested your changes:
Tested locally and also on testworld-2-0

Checklist:

- [ ] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

* Closes #0000

### CODE REVIEW

**Review by georgeee — APPROVED** [review]

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

File: src/app/archive/lib/processor.ml [hunk 1]
```diff
@@ -1417,20 +1421,119 @@ module Zkapp_events = struct
...
+        let field_list_list =
+          List.map events ~f:(fun field_array ->
+              Array.map field_array ~f:Pickles.Backend.Tick.Field.to_string
+              |> Array.to_list )
+        in
+        let fields = field_list_list |> List.concat in
+        let%bind field_id_list_list =
+          if not @@ List.is_empty fields then
+            let field_insert =
+              sprintf
```
Comment (georgeee) [inline src/app/archive/lib/processor.ml:1458]: Do you think we can generalize the pattern and extract it as a separate function to reduce code duplication?

I'm taking about `insert on conflict do nothing` paired with select of `some_field, id` with conversion of it to map preserving the initial order.

**Review by georgeee — CHANGES_REQUESTED** [review]

File: src/lib/mina_caqti/mina_caqti.ml [hunk 2]
```diff
@@ -287,15 +287,19 @@ let select_insert_into_cols ~(select : string * 'select Caqti_type.t)
     @@ select_cols ~select:(fst select) ~table_name ?tannot ~cols:(fst cols) ()
     )
     value
-  >>= function
```
[the remark this answers is not in the document]
Comment (ghost-not-in-the-shell) [reply src/lib/mina_caqti/mina_caqti.ml:290]: Ah. I changed it to add some more logs. Let me revert those changes.

File: src/app/archive/lib/processor.ml [hunk 3]
```diff
@@ -1417,20 +1421,119 @@ module Zkapp_events = struct
...
+            in
+            let%map field_map =
+              Conn.collect_list
+                (Caqti_request.collect Caqti_type.unit
+                   Caqti_type.(tup2 string int)
+                   field_search )
+                ()
+              >>| String.Map.of_alist_exn
+            in
+            let field_id_list_list =
```
[the remark this answers is not in the document]
Comment (ghost-not-in-the-shell) [reply src/app/archive/lib/processor.ml:1481]: I added a let binding here to make the code more readable. Just to give a name to the result of that computation.

File: src/app/archive/lib/processor.ml (hunk 1 above)
Comment (ghost-not-in-the-shell) [reply src/app/archive/lib/processor.ml:1458]: Yeah. Let me add a function in mina_caqti for that.

File: src/lib/mina_caqti/mina_caqti.ml [hunk 4]
```diff
@@ -287,15 +287,19 @@ let select_insert_into_cols ~(select : string * 'select Caqti_type.t)
...
+  match res with
   | Some id ->
       return id
   | None ->
-      Conn.find
-        ( Caqti_request.find (snd cols) (snd select)
-        @@ insert_into_cols ~returning:(fst select) ~table_name ?tannot
-             ~cols:(fst cols) () )
-        value
+      let%map res =
```
[the remark this answers is not in the document]
Comment (ghost-not-in-the-shell) [reply src/lib/mina_caqti/mina_caqti.ml:295]: Again this change was for logging. Let me undo that.

File: src/app/archive/lib/processor.ml (hunk 1 above)
Comment (ghost-not-in-the-shell) [reply src/app/archive/lib/processor.ml:1458]: Nevermind, I got most part of the refactor done. I left out the conversion from value to string.

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Review by georgeee — APPROVED** [review]

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Review by deepthiskumar — DISMISSED** [review]
Waiting for rampup8 release before landing this

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
!ci-nightly-me

**Comment by deepthiskumar** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Review by joaosreis — APPROVED** [review]

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Review by mitschabaude — APPROVED** [review]

**Comment by deepthiskumar** [comment]
!approved-for-mainnet

**Pull request merged**